Userland/Games/Solitaire/Game.cpp

```diff
@@ -482,6 +497,72 @@ void Game::paint_event(GUI::PaintEvent& event)
     }
 }
 
+void Game::undo_save_move(CardStack& from, CardStack& to, NonnullRefPtrVector<Card> moved_cards)
+{
+    m_last_move.type = LastMove::Type::MoveCards;
+    m_last_move.from = &from;
+    m_last_move.cards = moved_cards;
+    m_last_move.to = &to;
+    on_undo_available(true);
+}
+
+void Game::undo_save_flip(Card& card)
+{
+    NonnullRefPtrVector<Card> cards;
+    cards.append(card);
+    m_last_move.type = LastMove::Type::FlipCard;
+    m_last_move.cards = cards;
+    on_undo_available(true);
+}
+
+void Game::undo_perform()
+{
+    if (!m_last_move.from)
+        return;
+
+    if (m_last_move.type == LastMove::Type::FlipCard) {
+        m_last_move.cards.at(0).set_upside_down(true);
+        on_undo_available(false);
+        invalidate_layout();
+        return;
+    }
+
+    if (m_last_move.from->type() == CardStack::Type::Play && m_mode == Mode::SingleCardDraw) {
+        auto& waste = stack(Waste);
+        if (!m_last_move.from->is_empty())
+            waste.push(m_last_move.from->pop());
+    }
+
+    for (auto& to_intersect : m_last_move.cards) {
+        mark_intersecting_stacks_dirty(to_intersect);
+        m_last_move.from->push(to_intersect);
+        m_last_move.to->pop();
+    }
+
+    if (m_last_move.from->type() == CardStack::Type::Stock) {
+        auto& waste = this->stack(Waste);
+        auto& play = this->stack(Play);
+        NonnullRefPtrVector<Card> cards_popped;
+        for (size_t i = 0; i < m_last_move.cards.size(); i++) {
+            if (!waste.is_empty()) {
+                auto card = waste.pop();
+                cards_popped.prepend(card);
+            }
+        }
+        for (auto& card : cards_popped) {
+            m_focused_cards.append(card);
+            play.push(move(card));
+        }
+    }
+
+    score_move(*m_last_move.from, *m_last_move.to, true);
+
+    m_last_move.from = nullptr;
+    m_last_move.to = nullptr;
+    on_undo_available(false);
```

Comment:
Please null-check this before invoking it:
```suggestion
    if (on_undo_available)
        on_undo_available(false);
```

Reply:
(+ same in multiple other places)